profiles/killbill/src/main/java/org/killbill/billing/server/notifications/PushNotificationListener.java

```diff
@@ -112,13 +137,104 @@ public Response onCompleted(final Response response) throws Exception {
                         }
                     });
             response = futureStatus.get(timeoutSec, TimeUnit.SECONDS);
+        } catch (final TimeoutException toe) {
+            saveRetryPushNotificationInQueue(tenantId, url, body);
+            return false;
         } catch (final Exception e) {
             log.warn("Failed to push notification url='{}', tenantId='{}'", url, tenantId, e);
             return false;
         }
+
+        if ((response.getStatusCode() == 301) || (response.getStatusCode() == 500)) {
+            saveRetryPushNotificationInQueue(tenantId, url, body);
+            return false;
+        }
+
         return response.getStatusCode() >= 200 && response.getStatusCode() < 300;
     }
 
+    public void resendPushNotification(final PushNotificationKey key, final InternalCallContext callContext) throws JsonProcessingException {
+
```

Comment:
Most of the code to send the notification is the same one that was used when sending initially. Is there a way to reuse that code instead to avoid duplication?